File: src/actions/system_upgrade.go
package actions

import (
	"enva/libs"
	"enva/services"
	"strings"
)

// SystemUpgradeAction runs system upgrade
type SystemUpgradeAction struct {
	*BaseAction
}

func NewSystemUpgradeAction(sshService *services.SSHService, aptService *services.APTService, pctService *services.PCTService, containerID *string, cfg *libs.LabConfig, containerCfg *libs.ContainerConfig) Action {
	return &SystemUpgradeAction{
		BaseAction: &BaseAction{
			SSHService:   sshService,
			APTService:   aptService,
			PCTService:   pctService,
			ContainerID: containerID,
			Cfg:          cfg,
			ContainerCfg: containerCfg,
		},
	}
}

func (a *SystemUpgradeAction) Description() string {
	return "system upgrade"
}

func (a *SystemUpgradeAction) Execute() bool {
	if a.APTService == nil {
		libs.GetLogger("system_upgrade").Printf("APT service not initialized")
		return false
	}

	// Run apt update
	output, exitCode := a.APTService.Update()
	if exitCode == nil || *exitCode != 0 {
		libs.GetLogger("system_upgrade").Printf("apt update failed")
		return false
	}

	// Run distribution upgrade
	output, exitCode = a.APTService.DistUpgrade()
	if exitCode == nil || *exitCode != 0 {
		outputLower := strings.ToLower(output)
		successIndicators := strings.Contains(outputLower, "setting up") ||
			strings.Contains(outputLower, "processing triggers") ||
			strings.Contains(outputLower, "created symlink") ||
			strings.Contains(outputLower, "0 upgraded") ||
			strings.Contains(outputLower, "0 newly installed")
		if !successIndicators {
			libs.GetLogger("system_upgrade").Printf("distribution upgrade failed: %s", output)
			return false
		}
	}
	return true
}

File: csharp-enva/Actions/system_upgrade.cs
using Enva.Libs;
using Enva.Services;

namespace Enva.Actions;

public class SystemUpgradeAction : BaseAction, IAction
{
    public SystemUpgradeAction(SSHService? sshService, APTService? aptService, PCTService? pctService, string? containerID, LabConfig? cfg, ContainerConfig? containerCfg)
    {
        SSHService = sshService;
        APTService = aptService;
        PCTService = pctService;
        ContainerID = containerID;
        Cfg = cfg;
        ContainerCfg = containerCfg;
    }

    public override string Description()
    {
        return "system upgrade";
    }

    public bool Execute()
    {
        if (APTService == null)
        {
            Logger.GetLogger("system_upgrade").Printf("APT service not initialized");
            return false;
        }

        // Run apt update
        var (output, exitCode) = APTService.Update();
        if (!exitCode.HasValue || exitCode.Value != 0)
        {
            Logger.GetLogger("system_upgrade").Printf("apt update failed");
            return false;
        }

        // Run distribution upgrade
        (output, exitCode) = APTService.DistUpgrade();
        if (!exitCode.HasValue || exitCode.Value != 0)
        {
            string outputLower = output.ToLower();
            bool successIndicators = outputLower.Contains("setting up") ||
                outputLower.Contains("processing triggers") ||
                outputLower.Contains("created symlink") ||
                outputLower.Contains("0 upgraded") ||
                outputLower.Contains("0 newly installed");
            if (!successIndicators)
            {
                Logger.GetLogger("system_upgrade").Printf("distribution upgrade failed: {0}", output);
                return false;
            }
        }
        return true;
    }
}

public static class SystemUpgradeActionFactory
{
    public static IAction NewSystemUpgradeAction(SSHService? sshService, APTService? aptService, PCTService? pctService, string? containerID, LabConfig? cfg, ContainerConfig? containerCfg)
    {
        return new SystemUpgradeAction(sshService, aptService, pctService, containerID, cfg, containerCfg);
    }
}
